Source/WebSocket/hcwebsocket.cpp

```diff
@@ -468,28 +482,40 @@ void HC_WEBSOCKET::CloseFunc(
 )
 {
     UNREFERENCED_PARAMETER(context);
+
+    bool shouldNotClose { false };
+
     {
         std::lock_guard<std::recursive_mutex> lock{ websocket->m_mutex };
 
+        shouldNotClose = (websocket->m_state == State::Connecting || websocket->m_state == State::Disconnected);
```

Comment:
>shouldNotClose = (websocket->m_state == State::Connecting || websocket->m_state == State::Disconnected); [](start = 8, length = 104)

Did you test this on other platforms?  I assume the expectation is that if connect fails, the provider should complete the connect XAsyncBlock & not call the closeFunc, which seems reasonable.  Are we sure all our providers are actually doing this though?  I think we need to make it 100% clear what the expectations of the provider are is essential as it seems we have some inconsistency across the various providers here

In any case, I think this would read a bit clearer if you just broke this up into separate cases and commented on each:

switch(websocket->m_state)
{
case State::Connecting:
  // provider error: providers should complete connect async context rather than calling closeFunc if a connect attempt fails
  return;
case State::Disconnected:
  // provider error: providers should call closeFunc exactly once per successful attempt
  return;
case State::Initial:
  // ...
 return;
case State::Connected:
case State::Disconnecting:
  // Expected. Invoke client callback and update our state
  ....
}

This would also help us catch any inconsistency in provider implementation more easily.